src/core/qgsdataitem.h

```diff
@@ -143,6 +143,18 @@ class CORE_EXPORT QgsDataItem : public QObject
      */
     virtual QList<QAction *> actions() { return QList<QAction *>(); }
 
+    /** Returns the list of menus available for this item. This is usually used for the popup menu on right-clicking
+     * the item. Subclasses should override this to provide actions. Subclasses should ensure that ownership of
+     * created menus is correctly handled by parenting them to the specified parent widget.
+     * \param parent a parent widget of the menu
+     * \returns list of menus
+     */
+    virtual QList<QMenu *> menus( QWidget *parent )
+    {
```

Comment:
Better move the implementation to cpp files (super-easy job with the context menu refactor of QtCreator ;) )